### PR TITLE
Feature/sortable reward area

### LINKED ISSUE #9630: Pinned Items should be sortable
You should be able to sort all items pinned to your Rewards column using drag-and-drop.

### PR DESCRIPTION
**Pull request by kipraske**
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #9630

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)

This pull request is to let the user sort the square pinnable habitica rewards at the bottom of the 4th column using drag and drop. This sounded easier to do than I expected at the time, but those sorts of things are the best learning opportunities.

The code changes have three main components:
  * The frontend drag mechanics
  * The backend saving mechanics
  * Using the saved fields to get items in order like tasks

#### Frontend Changes

The frontend code is pretty straightforward but I am new to vue.js to please check this to make sure I didn't completely mess any best practices up. The frontend console and webpack are not currently complaining so that is a good sign.

The rewards area needed to be in a `<draggable>` component, and so I created one of those with the same class as the existing rewards container so we didn't mess up the CSS

I ran into an issue that the popovers still worked on hover on these items and made the drag and drop really annoying to use. So I added some code to disable the popovers on the reward items while you are in mid-drag.

#### Backend Changes

To save this information I made a new field on the user object called `user.pinnedItemsOrder`. I wanted to avoid making a new field on the user but because of the way that pinned items come from two different places (the user pinned items and the automatic seasonal items) and then are combined on the frontend it was the only way I could get it to work properly.

I created a new API endpoint to handle these changes and update this field.

I wrote this code so that `user.pinnedItemsOrder` doesn't need to be updated unless the user actually wants a custom ordering. If it is uninitialized you just get exactly what you get now: the items in the default ordering. If items are pinned or unpinned then it orders what we have ordered before and new items are just placed at the top of the list. The same is true for adding or removing new holiday items.

`user.pinnedItemsOrder` is keyed using the `item.path` variable since these items don't have the UUID consistently available for checking. The `item.path` is used in the API so should be unique though it is not programmatically guaranteed I suppose.

#### Retrieving Ordered Tasks

When you call `inAppRewards()` now to get the rewards for the frontend the code now checks the pinnedItemsOrder to retrieve them in the correct order. The logic is the exact same thing that is used for tasks and taskOrder for each of the columns.

The tricky part of all this code is when there is a different number of rewards out on the frontend than we have saved in our tracking order. But because we can use `inAppRewards` now to get the tasks in the order that we expect, if we ever need to reset `user.pinnedItemsOrder` before we do any moving or saving we can just reset `user.pinnedItemsOrder = inAppRewards` (see line 1982 in website/server/controllers/api-v3/user.js)

#### Current Status:
The current status is that I think I need help testing and deciding what sorts of automated tests are needed for this code. I have no idea how to add or remove the seasonal pinned items from the code and we should make sure that doing that doesn't break anything. I did some tests to simulate this, but it wasn't the real-deal.

[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID: f5d12242-5b66-471d-bfc2-6c3358861d9c

### CODE REVIEW

**Comment by negue** [comment]
Code and the feature in the website are LGTM 👍

**Comment by negue** [comment]
I think we'd still need some tests, unit and api 

@kipraske could you add some ?

If you need help just ask :)

**Comment by kipraske** [comment]
Sounds good. I am happy to work on writing up those tests. I am not particularly familiar with mocha so it will take me a little bit to get up to speed, but I am glad to learn it.

**Comment by kipraske** [comment]
@negue ~ So it turns out that I do need a little help. In particular the tricky part of my code is dealing with the seasonal items when they come into the pinned rewards area and also when they leave. So the main problem is: How can I activate a set of seasonal items for the purpose of testing? After all, if the seasonal items exist they will find their way into `user.pinnedItemsOrder` whether I know what they are in advance or not.

The good news is writing this post, I think I have figured out a way to get my tests to consider these items, but I will need to pull in additional functions to get the test set up which I was trying to avoid, but I suppose it is an "integration test" so a little bit of that is okay I hope. So, if there is a way to activate/deactivate seasonal items that would be best and if not I have a plan B.

**Comment by negue** [comment]
Sorry for the long response time.

You can use `New Years` example to test: 
https://github.com/HabitRPG/habitica/blob/d57c525fab1fd0ee8d5dd6f3d16179770c8bb107/website/common/script/libs/shops-seasonal.config.js

maybe you also need to change the dates on 
- `contants.js` -> `EVENTS` of `winter2018`
- `shops-seasonal.config.js` -> `dateRange`


Looking at the code it should work with `officalPinnedItems` without any change.

**Comment by kipraske** [comment]
Well, I wrote two tests so far. I was in the middle of writing the last one that I thought we needed and it failed! So I guess there is a bug when you try to move an item that you don't actually have pinned you end up with null values in your order which is not so great. It should error out when you don't actually have it.

So I guess I will work on fixing that bug next and then let you know when it is ready for review again ~

**Comment by kipraske** [comment]
Okay I am happy with the API tests now, so I went on to the last two "easy" tests of my changes to `common/script/libs/inAppRewards.js`. I made a file in `tests/common/libs` for it. I can't seem to get it to run at all with this error:

```
 1) "before all" hook
  2) "after all" hook

  0 passing (8s)
  2 failing

  1)  "before all" hook:
     Error: Timeout of 8000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.


  2)  "after all" hook:
     TypeError: Cannot read property 'dropDatabase' of undefined
      at Context.<anonymous> (test/helpers/mongo.js:122:3)
```

 I tried to copy the same format I did for the API call tests but something is different here that I can't seem to track down. So I think I am officially stuck now and will need some help. The good news is that I am pretty close to being done now ~

**Comment by paglias** [comment]
I'm restarting the tests, they fail sometime

**Comment by kipraske** [comment]
To reiterate, there is a problem with my inapprewards.js tests, I am not sure how to fix them. There may also be problems in the tests but it is failing before the tests are even run...

**Comment by paglias** [comment]
Can you fix the linting problems listed here https://travis-ci.org/HabitRPG/habitica/jobs/354045454?

File: test/common/libs/inAppRewards.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+      'food.Saddle',
+      'gear.flat.shield_warrior_1',
+      'potion',
+    ];
+
+    // For this test put seasonal items at the end so they stay out of the way
+    testPinnedItemsOrder = testPinnedItemsOrder.concat(officialPinnedItemPaths);
+  });
+
+  it.only('returns the pinned items in the correct order', async () => {
```
Comment (paglias) [inline test/common/libs/inAppRewards.js]: .only should be removed

**Comment by kipraske** [comment]
Okay I updated the file to remove the code style errors and changed the tests that they probably will work. I still cannot run the tests in `tests/common/libs/inAppRewards.js` so they may have errors in them still, so any help would be appreciated.

**Comment by negue** [comment]
I tested it again, the feature works, but I can't run the common test either, the `before all` and `after all` error is always displayed. I can't figure out what is wrong with that test, because if I delete the file, it works just fine

**Comment by kipraske** [comment]
It seems like there is an async call somewhere and the test can't figure out that the call has completed for some reason. This seems to keep happening if I remove all of the "async/await" keywords, and it probably happens when I call generateUser. The strange thing is that I did the exact same thing in the other test file. Maybe there is some configuration of those files that assume we don't use async calls?

Well, worse case scenario we remove those tests? I just wanted them to make sure that everything didn't get strange when we changed from having seasonal items or not ~ For now we can wait to see if anyone else has any ideas about this error.

**Comment by paglias** [comment]
@kipraske @negue I was able to give a better look at this and it looks like the problem comes from the fact that you're using helpers written for integration tests (with the whole server running) in common tests where instead the idea is to test each module / function in isolation.

Taking an example from test/common/libs/updateStore, instead of using the async function `generateUser` from the integration tests helpers you should use the one from the common helpers which simply returns an user object without involving the database (nothing gets saved, it's simply kept in memory) https://github.com/HabitRPG/habitica/blob/develop/test/common/libs/updateStore.js#L2

There was an other error with the inAppRewards function not being passed the user.

This is the correct version of the test file 

```
matteo@macbook~/dev/habitrpg/habitica$ git diff
diff --git a/test/common/libs/inAppRewards.js b/test/common/libs/inAppRewards.js
index 5f7d4ad0a..d7634c72b 100644
--- a/test/common/libs/inAppRewards.js
+++ b/test/common/libs/inAppRewards.js
@@ -1,6 +1,6 @@
 import {
   generateUser,
-} from '../../helpers/api-integration/v3';
+} from '../../helpers/common.helper';
 import getOfficialPinnedItems from '../../../website/common/script/libs/getOfficialPinnedItems.js';
 import inAppRewards from '../../../website/common/script/libs/inAppRewards';

@@ -11,8 +11,8 @@ describe('inAppRewards', () => {
   let testPinnedItems;
   let testPinnedItemsOrder;

-  beforeEach(async () => {
-    user = await generateUser();
+  beforeEach(() => {
+    user = generateUser();
     officialPinnedItems = getOfficialPinnedItems(user);

     officialPinnedItemPaths = [];
@@ -52,18 +52,17 @@ describe('inAppRewards', () => {
     testPinnedItemsOrder = testPinnedItemsOrder.concat(officialPinnedItemPaths);
   });

-  it('returns the pinned items in the correct order', async () => {
-    await user.update({
-      pinnedItems: testPinnedItems,
-      pinnedItemsOrder: testPinnedItemsOrder,
-    });
-    await user.sync();
-    let result = inAppRewards();
+  it('returns the pinned items in the correct order', () => {
+    user.pinnedItems = testPinnedItems;
+    user.pinnedItemsOrder = testPinnedItemsOrder;
+
+    let result = inAppRewards(user);
+
     expect(result[2].path).to.eql('armoire');
     expect(result[9].path).to.eql('potion');
   });

-  it('does not return seasonal items which have been unpinned', async () => {
+  it('does not return seasonal items which have been unpinned', () => {
     if (officialPinnedItems.length === 0) {
       return; // if no seasonal items, this test is not applicable
     }
@@ -74,15 +73,12 @@ describe('inAppRewards', () => {
       { type: testUnpinnedItem.type, path: testUnpinnedPath},
     ];

-    await user.update({
-      pinnedItems: testPinnedItems,
-      pinnedItemsOrder: testPinnedItemsOrder,
-      unpinnedItems: testUnpinnedItems,
-    });
-    await user.sync();
+    user.pinnedItems = testPinnedItems;
+    user.pinnedItemsOrder = testPinnedItemsOrder;
+    user.unpinnedItems = testUnpinnedItems;

-    let result = inAppRewards();
+    let result = inAppRewards(user);
     let itemPaths = result.map(item => item.path);
-    expect(itemPaths).to.not.inlude(testUnpinnedPath);
+    expect(itemPaths).to.not.include(testUnpinnedPath);
   });
 });
(END)
```

Also in text format instead of git diff so it's easier to apply to your changes

```
import {
  generateUser,
} from '../../helpers/common.helper';
import getOfficialPinnedItems from '../../../website/common/script/libs/getOfficialPinnedItems.js';
import inAppRewards from '../../../website/common/script/libs/inAppRewards';

describe('inAppRewards', () => {
  let user;
  let officialPinnedItems;
  let officialPinnedItemPaths;
  let testPinnedItems;
  let testPinnedItemsOrder;

  beforeEach(() => {
    user = generateUser();
    officialPinnedItems = getOfficialPinnedItems(user);

    officialPinnedItemPaths = [];
    // officialPinnedItems are returned in { type: ..., path:... } format but we just need the paths for testPinnedItemsOrder
    if (officialPinnedItems.length > 0) {
      officialPinnedItemPaths = officialPinnedItems.map(item => item.path);
    }

    testPinnedItems = [
      { type: 'armoire', path: 'armoire' },
      { type: 'potion', path: 'potion' },
      { type: 'marketGear', path: 'gear.flat.weapon_warrior_1' },
      { type: 'marketGear', path: 'gear.flat.head_warrior_1' },
      { type: 'marketGear', path: 'gear.flat.armor_warrior_1' },
      { type: 'hatchingPotions', path: 'hatchingPotions.Golden' },
      { type: 'marketGear', path: 'gear.flat.shield_warrior_1' },
      { type: 'card', path: 'cardTypes.greeting' },
      { type: 'potion', path: 'hatchingPotions.Golden' },
      { type: 'card', path: 'cardTypes.thankyou' },
      { type: 'food', path: 'food.Saddle' },
    ];

    testPinnedItemsOrder = [
      'hatchingPotions.Golden',
      'cardTypes.greeting',
      'armoire',
      'gear.flat.weapon_warrior_1',
      'gear.flat.head_warrior_1',
      'cardTypes.thankyou',
      'gear.flat.armor_warrior_1',
      'food.Saddle',
      'gear.flat.shield_warrior_1',
      'potion',
    ];

    // For this test put seasonal items at the end so they stay out of the way
    testPinnedItemsOrder = testPinnedItemsOrder.concat(officialPinnedItemPaths);
  });

  it('returns the pinned items in the correct order', () => {
    user.pinnedItems = testPinnedItems;
    user.pinnedItemsOrder = testPinnedItemsOrder;

    let result = inAppRewards(user);

    expect(result[2].path).to.eql('armoire');
    expect(result[9].path).to.eql('potion');
  });

  it('does not return seasonal items which have been unpinned', () => {
    if (officialPinnedItems.length === 0) {
      return; // if no seasonal items, this test is not applicable
    }

    let testUnpinnedItem = officialPinnedItems[0];
    let testUnpinnedPath = testUnpinnedItem.path;
    let testUnpinnedItems = [
      { type: testUnpinnedItem.type, path: testUnpinnedPath},
    ];

    user.pinnedItems = testPinnedItems;
    user.pinnedItemsOrder = testPinnedItemsOrder;
    user.unpinnedItems = testUnpinnedItems;

    let result = inAppRewards(user);
    let itemPaths = result.map(item => item.path);
    expect(itemPaths).to.not.include(testUnpinnedPath);
  });
});
```

Let me know if you have any other problem and sorry for the delay

**Comment by kipraske** [comment]
I applied your fixes @paglias and they work great. The tests appear to be passing properly and everything still works. Thank you so much for your help, it didn't occur to me that I could get the job done without updating the user in the database somewhere. But of course we can do it this way, it is a unit test of the function 👍 .

So I grabbed the most recent upstream develop branch to make sure there were no merge conflicts, and there appears to be quite a few failing tests that seemingly have nothing to do with this pull request.

I ran the tests again locally on just `develop` and see that these are failing on the habitica develop branch upstream.

**Comment by paglias** [comment]
Thanks for this @kipraske , I'm going to bump you two tier 2 on Habitica since this is a significant contribution!

**Pull request merged**